cpp/src/centrality/betweenness_centrality_impl.cuh

```diff
@@ -275,39 +277,79 @@ void accumulate_vertex_results(
+      do_expensive_check);
+    
+    // Create vertex list using thrust::copy_if
+    rmm::device_uvector<vertex_t> vertices_at_distance_d_minus_1(num_vertices_at_distance, handle.get_stream());
+    
+    if (num_vertices_at_distance > 0) {
+      auto v_first = graph_view.local_vertex_partition_range_first();
+      thrust::copy_if(
+        handle.get_thrust_policy(),
+        thrust::make_counting_iterator(v_first),
+        thrust::make_counting_iterator(graph_view.local_vertex_partition_range_last()),
+        vertices_at_distance_d_minus_1.begin(),
+        [distances = distances.begin(), d, v_first] __device__(vertex_t vertex) {
+          auto offset = vertex - v_first;
+          return distances[offset] == (d - 1);
+        });
+    }
+    
+    // Count edges that will be processed (where dst at distance d and src at distance d-1)
+    auto num_edges_to_process = count_if_e(
       handle,
       graph_view,
       src_properties.view(),
       dst_properties.view(),
-      cugraph::edge_dummy_property_t{}.view(),
+      edge_dummy_property_t{}.view(),
       [d] __device__(auto, auto, auto src_props, auto dst_props, auto) {
-        if ((thrust::get<0>(dst_props) == d) && (thrust::get<0>(src_props) == (d - 1))) {
-          auto sigma_v = static_cast<weight_t>(thrust::get<1>(src_props));
-          auto sigma_w = static_cast<weight_t>(thrust::get<1>(dst_props));
-          auto delta_w = thrust::get<2>(dst_props);
-
-          return (sigma_v / sigma_w) * (1 + delta_w);
-        } else {
-          return weight_t{0};
-        }
+        return ((thrust::get<0>(dst_props) == d) && (thrust::get<0>(src_props) == (d - 1)));
       },
-      weight_t{0},
-      reduce_op::plus<weight_t>{},
-      deltas.begin(),
       do_expensive_check);
-
+    
+    if (vertices_at_distance_d_minus_1.size() > 0) {
+      // Create key_bucket_t from the vertex list
+      key_bucket_t<vertex_t, void, multi_gpu, true> vertex_list(
+        handle, std::move(vertices_at_distance_d_minus_1));
+      
+      // Use vertex list approach - only process vertices at distance d-1
+      per_v_transform_reduce_outgoing_e(
+        handle,
+        graph_view,
+        vertex_list,
+        src_properties.view(),
+        dst_properties.view(),
+        cugraph::edge_dummy_property_t{}.view(),
+        [d] __device__(auto, auto, auto src_props, auto dst_props, auto) {
+          if ((thrust::get<0>(dst_props) == d) && (thrust::get<0>(src_props) == (d - 1))) {
+            auto sigma_v = static_cast<weight_t>(thrust::get<1>(src_props));
+            auto sigma_w = static_cast<weight_t>(thrust::get<1>(dst_props));
+            auto delta_w = thrust::get<2>(dst_props);
+            
+            return (sigma_v / sigma_w) * (1 + delta_w);
+          } else {
+            return weight_t{0};
+          }
+        },
+        weight_t{0},
+        reduce_op::plus<weight_t>{},
+        deltas.begin(),
+        do_expensive_check);
+    }
```

Comment:
See
```
 * @param vertex_value_output_first Iterator pointing to the (tagged-)vertex property variables for
 * the first (inclusive) (tagged-)vertex in @p key_list. `vertex_value_output_last` (exclusive) is
 * deduced as @p vertex_value_output_first + @p key_list.size().
```
https://github.com/rapidsai/cugraph/blob/branch-25.08/cpp/src/prims/per_v_transform_reduce_incoming_outgoing_e.cuh#L362

Here, the output array size should be the size of the vertex_list (not the number of local vertices), and the first output value is for the first vertex in the vertex list.

Reply:
Something minor but,

```
auto v_first = graph_view.local_vertex_partition_range_first();
...
[centralities = centralities.data(), v_first]
```

can just be
```
[centrealities = centralities.data(), v_first = graph_view.local_vertex_partition_range_first()]
```

as this minimizes the scope of v_first.